pow/ethereumj/src/main/java/org/ethereum/beacon/pow/EthereumJDepositContract.java

```diff
@@ -0,0 +1,263 @@
+      if (DEPOSIT_EVENT_NAME.equals(invocation.function.name)) {
+        newDeposit(createDepositEventData(invocation), block.getHash());
+      } else if (CHAIN_START_EVENT_NAME.equals(invocation.function.name)) {
+        chainStart((byte[]) invocation.args[0], (byte[]) invocation.args[1], block.getHash());
+      } else {
+        throw new IllegalStateException("Invalid event from the contract: " + invocation);
+      }
+    }
+  }
+
+  private DepositEventData createDepositEventData(Invocation depositEvent) {
+    Object[] merkle_branch_obj = (Object[]) depositEvent.args[3];
+    byte[][] merkle_branch_arr = new byte[merkle_branch_obj.length][];
+    for (int i = 0; i < merkle_branch_obj.length; i++) {
+      merkle_branch_arr[i] = (byte[]) merkle_branch_obj[i];
+    }
+
+    return new DepositEventData(
+        (byte[]) depositEvent.args[0],
+        (byte[]) depositEvent.args[1],
+        (byte[]) depositEvent.args[2],
+        merkle_branch_arr);
+  }
+
+  private List<Invocation> getContractEvents(Block block) {
+    if (!new Bloom(block.getLogBloom()).matches(contractAdderssBloom)) {
+      return Collections.emptyList();
+    }
+    List<Invocation> ret = new ArrayList<>();
+
+    List<TransactionReceipt> receipts = getBlockTransactionReceipts(block);
+    for (TransactionReceipt receipt : receipts) {
+      for (LogInfo logInfo : receipt.getLogInfoList()) {
+        if (Arrays.equals(logInfo.getAddress(), contractDeployAddress.getArrayUnsafe())) {
+          ret.add(contract.parseEvent(logInfo));
+        }
+      }
+    }
+
+    return ret;
+  }
+
+  @Override
+  protected boolean hasDepositRootImpl(byte[] blockHash, byte[] depositRoot) {
+    Block block = ethereum.getBlockchain().getBlockByHash(blockHash);
+    if (block == null) {
+      return false;
+    }
+    if (ethereum.getBlockchain().getBestBlock().getNumber() - block.getNumber() < getDistanceFromHead()) {
+      return false;
+    }
+
+    return getContractEvents(block).stream()
+        .filter(invocation -> DEPOSIT_EVENT_NAME.equals(invocation.function.name))
+        .anyMatch(invocation -> Arrays.equals((byte[]) invocation.args[0], depositRoot));
+  }
+
+  @Override
+  protected Pair<byte[], byte[]> getLatestBlockHashDepositRoot() {
+    long bestBlock = ethereum.getBlockchain().getBestBlock().getNumber() - getDistanceFromHead();
+    for(long blockNum = bestBlock; blockNum >= contractDeployBlock; blockNum--) {
+      Block block = ethereum.getBlockchain().getBlockByNumber(blockNum);
+      List<Invocation> contractEvents = getContractEvents(block);
+      Collections.reverse(contractEvents);
+      for (Invocation contractEvent : contractEvents) {
+        if (CHAIN_START_EVENT_NAME.equals(contractEvent.function.name)) {
+          return Pair.with(block.getHash(), (byte[]) contractEvent.args[0]);
+        } else {
+          return Pair.with(block.getHash(), (byte[]) contractEvent.args[0]);
+        }
+      }
+    }
+    return null;
```

Comment:
Why not an `Optional` here?

Reply:
Kinda 'old style' with byte arrays and nulls :)
Ok, will change this